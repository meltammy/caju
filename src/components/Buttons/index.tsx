import styled from "styled-components";

const Button = styled.button`
  outline: none;
  display: flex;
  align-items: center;
  border: none;
  border-radius: 36px;
  padding: 8px 32px;
  background-color: #64a98c;
  cursor: pointer;
  height: 56px;
  color: #fff;
  box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;
  font-size: 16px;
  font-weight: 600;
`;

export default Button;
