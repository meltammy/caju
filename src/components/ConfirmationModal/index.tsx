import { FC } from "react";
import {
  Description,
  ModalContainer,
  ModalOverlay,
  Title,
  ButtonContainer,
  StyledAsyncButton,
} from "./styles";
import Button from "../Buttons/Button";

interface ConfirmationModalProps {
  title: string;
  description: string;
  confirmationButtonLabel: string;
  isOpen: boolean;
  isLoading?: boolean;
  onClose: () => void;
  onConfirm: () => void;
}

export const confirmationModalTestId = "confirmation-modal";
const ConfirmationModal: FC<ConfirmationModalProps> = ({
  title,
  description,
  confirmationButtonLabel,
  isOpen,
  isLoading,
  onClose,
  onConfirm,
}) => {
  return (
    <ModalOverlay isOpen={isOpen} data-testid={confirmationModalTestId}>
      <ModalContainer>
        <Title>{title}</Title>
        <Description dangerouslySetInnerHTML={{ __html: description }} />
        <ButtonContainer>
          <Button onClick={onClose} disabled={isLoading}>
            Cancelar
          </Button>

          <StyledAsyncButton
            isConfirmation
            onClick={onConfirm}
            isLoading={isLoading}
          >
            {confirmationButtonLabel}
          </StyledAsyncButton>
        </ButtonContainer>
      </ModalContainer>
    </ModalOverlay>
  );
};

export default ConfirmationModal;
