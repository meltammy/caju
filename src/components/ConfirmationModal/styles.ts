import styled from "styled-components";

export const ModalOverlay = styled.div<{ $isOpen: boolean }>`
  display: ${(props) => (props.$isOpen ? "flex" : "none")};
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(0, 0, 0, 0.5);
  justify-content: center;
  align-items: center;
`;

export const ModalContainer = styled.div`
  background: white;
  border-radius: 8px;
  padding: 20px;
  min-width: 500px;
  max-width: 95vw;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
  z-index: 999;

  @media screen and (max-width: 1200px) {
    min-width: 50vw;
    margin: 1rem;
  }
`;

export const Title = styled.h2`
  margin: 0;
`;

export const Description = styled.p`
  margin: 10px 0 20px;
`;

export const ButtonContainer = styled.div`
  display: flex;
  gap: 1rem;

  @media screen and (max-width: 1200px) {
    flex-direction: column-reverse;
  }

  button {
    width: 100%;
    place-content: center;
  }
`;
