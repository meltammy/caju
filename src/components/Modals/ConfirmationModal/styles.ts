import styled from "styled-components";

export const ModalContainer = styled.div`
  min-width: 32rem;
  max-width: 95vw;
  z-index: 999;
  padding: 1.5rem;
  background: white;
  border-radius: 0.5rem;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.desktop}) {
    min-width: 50vw;
    margin: 1rem;
  }
`;

export const Title = styled.h2`
  margin: 0;
`;

export const Description = styled.p`
  margin: 10px 0 1.5rem;
`;

export const ButtonContainer = styled.div`
  display: flex;
  gap: 1rem;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.desktop}) {
    flex-direction: column-reverse;
  }
`;
