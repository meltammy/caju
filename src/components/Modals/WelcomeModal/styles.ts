import styled, { keyframes } from "styled-components";
import Button from "~/components/Buttons/Button";

const coming = keyframes`
  0% {
    bottom: -73rem;
  }

  100% {
    bottom: 0;
  }
`;

export const ModalContainer = styled.div`
  position: relative;
  z-index: 999;

  width: 52rem;
  height: 90%;
  max-height: 30rem;
  background: white;
  border-radius: 0.5rem;
  padding: 1.5rem;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);

  b {
    color: ${({ theme }) => theme.colors.primary};
  }

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.desktop}) {
    position: absolute;
    width: 100%;
    min-width: 50vw;
    max-height: 36rem;
    margin-top: auto;

    display: grid;
    grid-template-rows: 1fr 10fr 1fr;

    border-bottom-right-radius: 0;
    border-bottom-left-radius: 0;

    animation: ${coming} 1s ease-in-out forwards;
  }
`;

export const StyledButton = styled(Button)`
  width: 60%;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.desktop}) {
    width: 100%;
  }
`;

export const Title = styled.h1`
  margin: 0;
  font-size: 3rem;
  color: ${({ theme }) => theme.colors.primary};

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.desktop}) {
    margin: 0% 0 2rem 0;
    font-size: 2rem;
    text-align: center;
  }
`;

export const DescriptionContainer = styled.div``;

export const Description = styled.p`
  font-size: 1rem;
  max-width: 60%;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.desktop}) {
    max-width: 100%;
    margin: 0 0 1rem 0;
    font-size: 0.875rem;
  }
`;

export const Illustration = styled.img`
  position: absolute;
  right: 5%;
  bottom: 5%;
  max-height: 60%;
  z-index: -1;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.desktop}) {
    max-height: 43%;
  }
`;
