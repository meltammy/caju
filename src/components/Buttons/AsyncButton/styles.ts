import styled from "styled-components";
import { Spinner } from "~/components/Spinner";

export const StyledSpinner = styled(Spinner)`
  width: 5px;
  height: 5px;
  border-width: 3px;
`;
