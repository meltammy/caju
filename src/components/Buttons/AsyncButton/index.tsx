import { ComponentProps } from "react";
import Button from "../Button";
import { StyledSpinner } from "./styles";

type Props = {
  isLoading?: boolean;
} & ComponentProps<typeof Button>;

export const asyncButtonTestId = "async-button";
export function AsyncButton({ isLoading, children, disabled, ...rest }: Props) {
  return (
    <Button
      {...rest}
      disabled={isLoading || disabled}
      data-testid={asyncButtonTestId}
    >
      {isLoading ? <StyledSpinner /> : children}
    </Button>
  );
}
