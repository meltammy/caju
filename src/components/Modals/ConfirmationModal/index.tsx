import { FC } from "react";
import * as S from "./styles";
import Button from "../../Buttons/Button";
import { AsyncButton } from "../../Buttons/AsyncButton";
import { ModalOverlay } from "../ModalOverlay";

interface ConfirmationModalProps {
  title: string;
  description: string;
  confirmationButtonLabel: string;
  isOpen: boolean;
  isLoading?: boolean;
  onClose: () => void;
  onConfirm: () => void;
  confirmationButtonColor?: string;
}

type Event = { stopPropagation: () => void };

export const confirmationModalTestId = "confirmation-modal";

const ConfirmationModal: FC<ConfirmationModalProps> = ({
  title,
  description,
  confirmationButtonLabel,
  confirmationButtonColor: confirmationButtonColor = "primary",
  isOpen,
  isLoading,
  onClose,
  onConfirm,
}) => {
  return (
    <ModalOverlay
      onClick={onClose}
      $isOpen={isOpen}
      data-testid={confirmationModalTestId}
    >
      <S.ModalContainer onClick={(event: Event) => event?.stopPropagation()}>
        <S.Title id="modal-title">{title}</S.Title>
        <S.Description dangerouslySetInnerHTML={{ __html: description }} />
        <S.ButtonContainer>
          <Button
            $variant="outline"
            color="grey"
            onClick={onClose}
            disabled={isLoading}
            id="cancel-button"
          >
            Cancelar
          </Button>

          <AsyncButton
            $variant="outline"
            color={confirmationButtonColor}
            onClick={onConfirm}
            isLoading={isLoading}
            id="confirm-button"
          >
            {confirmationButtonLabel}
          </AsyncButton>
        </S.ButtonContainer>
      </S.ModalContainer>
    </ModalOverlay>
  );
};

export default ConfirmationModal;
