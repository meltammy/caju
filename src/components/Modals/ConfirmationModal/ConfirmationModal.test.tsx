import "@testing-library/jest-dom";
import "jest-styled-components";

import { render, fireEvent, screen } from "@testing-library/react";
import ConfirmationModal, { confirmationModalTestId } from ".";
import { ThemeProvider } from "styled-components";
import { theme } from "~/../styles/theme";
import { asyncButtonTestId } from "../../Buttons/AsyncButton";

const title = "Confirm Action";
const description = "Are you sure you want to proceed?";
const cancelButtonLabel = "Cancelar";

const renderModal = (
  props: Partial<React.ComponentProps<typeof ConfirmationModal>> = {}
) => {
  const defaultProps = {
    title,
    description,
    confirmationButtonLabel: "Yes",
    isOpen: true,
    onClose: jest.fn(),
    onConfirm: jest.fn(),
    ...props,
  };

  return render(
    <ThemeProvider theme={theme}>
      <ConfirmationModal {...defaultProps} />
    </ThemeProvider>
  );
};

describe("ConfirmationModal component", () => {
  it("should render modal with title and description", () => {
    renderModal();

    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(description)).toBeInTheDocument();
  });

  it("should calls onClose when cancel button is clicked", () => {
    const onClose = jest.fn();
    renderModal({ onClose });

    fireEvent.click(screen.getByText(cancelButtonLabel));
    expect(onClose).toHaveBeenCalled();
  });

  it("should calls onConfirm when confirmation button is clicked", () => {
    const onConfirm = jest.fn();
    renderModal({ onConfirm });

    fireEvent.click(screen.getByText("Yes"));
    expect(onConfirm).toHaveBeenCalled();
  });

  it("should disables buttons when loading", () => {
    renderModal({ isLoading: true });

    const cancelButton = screen.getByRole("button", { name: /cancelar/i });
    const confirmButton = screen.getByTestId(asyncButtonTestId);

    expect(cancelButton).toBeDisabled();
    expect(confirmButton).toBeDisabled();
  });

  it("should is not visible when isOpen is false", () => {
    renderModal({ isOpen: false });

    const modal = screen.getByTestId(confirmationModalTestId);
    expect(modal).toHaveStyleRule("display", "none");
  });

  it("should match snapshot when open", () => {
    const { asFragment } = renderModal();
    expect(asFragment()).toMatchSnapshot();
  });
});
