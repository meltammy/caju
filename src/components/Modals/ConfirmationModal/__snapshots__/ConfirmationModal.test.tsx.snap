// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationModal component should match snapshot when open 1`] = `
<DocumentFragment>
  .c1 {
  background: white;
  border-radius: 0.5rem;
  padding: 1.5rem;
  min-width: 32rem;
  max-width: 95vw;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
  z-index: 999;
}

.c2 {
  margin: 0;
}

.c3 {
  margin: 10px 0 1.5rem;
}

.c4 {
  display: flex;
  gap: 1rem;
}

.c4 button {
  width: 100%;
  place-content: center;
  position: absolute;
}

.c5 {
  outline: none;
  display: flex;
  align-items: center;
  border: none;
  border-radius: 1rem;
  padding: 0.5rem 2rem;
  place-content: center;
  cursor: pointer;
  height: 3.5rem;
  box-shadow: rgba(149, 157, 165, 0.2) 0px 0.5rem 1.5rem;
  font-size: 1rem;
  font-weight: 600;
  text-wrap: nowrap;
  border: 2px solid grey;
  color: grey;
  background: transparent;
}

.c5:disabled {
  background-color: #ccc;
  opacity: 1;
}

.c6 {
  outline: none;
  display: flex;
  align-items: center;
  border: none;
  border-radius: 1rem;
  padding: 0.5rem 2rem;
  place-content: center;
  cursor: pointer;
  height: 3.5rem;
  box-shadow: rgba(149, 157, 165, 0.2) 0px 0.5rem 1.5rem;
  font-size: 1rem;
  font-weight: 600;
  text-wrap: nowrap;
  border: 2px solid primary;
  color: primary;
  background: transparent;
}

.c6:disabled {
  background-color: #ccc;
  opacity: 1;
}

.c0 {
  display: flex;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(0, 0, 0, 0.5);
  justify-content: center;
  align-items: center;
  z-index: 999;
}

@media screen and (max-width: ${({theme}) => theme.breakpoints.desktop}) {
  .c1 {
    min-width: 50vw;
    margin: 1rem;
  }
}

@media screen and (max-width: ${({theme}) => theme.breakpoints.desktop}) {
  .c4 {
    flex-direction: column-reverse;
  }
}

<div
    class="c0"
    data-testid="confirmation-modal"
  >
    <div
      class="c1"
    >
      <h2
        class="c2"
        id="modal-title"
      >
        Confirm Action
      </h2>
      <p
        class="c3"
      >
        Are you sure you want to proceed?
      </p>
      <div
        class="c4"
      >
        <button
          class="c5"
          color="grey"
          id="cancel-button"
        >
          Cancelar
        </button>
        <button
          class="c6"
          color="primary"
          data-testid="async-button"
          id="confirm-button"
        >
          Yes
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
